### PR TITLE
Require types to start with an uppercase letter

### LINKED ISSUE #1547: `path` type mentions `entity` which is not a real type
**Describe the bug**
The `Typecheck` module contains `Path -> [tyQ| (unit + int) -> ((int * int) + entity) -> cmd (unit + dir) |]`, but we do not have any built-in type named `entity` (not yet at least; see #455), so it is interpreted as a type variable.  Writing `path` at the prompt displays the type
```
∀ a. unit + int -> int * int + a -> cmd (unit + dir)
```
where you can see that the `entity` has been replaced with a generic quantified type variable `a`.  However, the `path` command is expecting a `text` value, so we can use this to cause a fatal crash by typing e.g. `path (inl ()) (inr 3)` (see the below screenshot).

**Expected behavior**
Presumably the type of `path` should be `(unit + int) -> ((int * int) + text) -> cmd (unit + dir)`.

**Screenshots**
path-type-bug

### PR DESCRIPTION
**Pull request by byorgey**
Closes #1547 , which was caused by misspelled/nonexistent types being interpreted as type variables.  In #1550 I proposed one solution to the problem, namely, to stop implicitly quantifying types and require explicit `forall` on any polymorphic type.  This PR represents an alternative solution: to keep implicit quantification but require all types to start with an uppercase letter, as in Haskell.  I think I like this one better but I'm open to feedback.

Specifically, with this PR:
- All built-in type constructors (`Int`, `Cmd`, `Unit`, etc.) must start with a capital letter
- Type variables:
    - Must start with a lowercase letter or underscore
    - May not be named a lowercase version of a type constructor, *e.g.* `int`
        - This is important so that old code with lowercase types is not silently accepted by parsing the former types as implicitly quantified type variables; even in cases where old code no longer typechecks, this will give better error messages.
- Term variables:
    - May start with upper- or lowercase
    - May be named lowercase versions of type names, *e.g.* `let f : Int -> Int = \int. int + 1` is fine

This PR obviously represents a bigger breaking change.  Once we merge this we might consider adding an option to `swarm format` to be able to parse old code with lowercase types and then reformat it with uppercase, to aid in porting code.

Once this is merged I will also regenerate the wiki pages that mention types.

Closes #1550.

### CODE REVIEW

**Review by xsebek — APPROVED** [review]
The types look more readable to me 👍

Just to check, uppercase is parsed as a concrete type or fails if it does not exist, right? Apologies if I overlooked a test for it.

**Comment by byorgey** [comment]
> Just to check, uppercase is parsed as a concrete type or fails if it does not exist, right? Apologies if I overlooked a test for it.

That is correct.  I will add some more tests.

**Comment by xsebek** [comment]
Btw. I think that if you `show`ed data from Swarm 0.4 and `read` and `print`ed them by 0.5, that would work as that tool.

It could look like:
```sh
swarm-0.4 format --out=Haskell old_code.sw | swarm-0.5 format --in=Haskell --stdin
```

**Comment by byorgey** [comment]
> Btw. I think that if you `show`ed data from Swarm 0.4 and `read` and `print`ed them by 0.5, that would work as that tool.

That's true, but (1) that would require adding `--out` and `--in` flags, (2) it would be kind of annoying to tell users that in order to upgrade they needed to keep around an old binary instead of just replacing that old binary with the newer one, as usual.

**Comment by byorgey** [comment]
@kostmo I've resurrected this PR.  When you get a chance, take a look and let me know what you think.

I was going to give `swarm format` the ability to convert code from the old format to the new.  I may still do that but I don't think it has to be in this PR.

File: src/swarm-lang/Swarm/Language/Parser/Lex.hs [hunk 1]
```diff
@@ -142,54 +149,83 @@ operatorChar = T.singleton <$> oneOf opChars
...
+-- | Names of types built into the language.
+primitiveTypeNames :: [Text]
+primitiveTypeNames = "Cmd" : baseTypeNames
+
+-- | List of keywords built into the language.
+keywords :: [Text]
+keywords = T.words "let in def end true false forall require requirements"
+
 -- | List of reserved words that cannot be used as variable names.
 reservedWords :: [Text]
```
Comment (kostmo) [inline src/swarm-lang/Swarm/Language/Parser/Lex.hs:165]: Make this a `Set` since `elem` is used in a number of places?

**Review by kostmo — APPROVED** [review]

File: src/swarm-tui/Swarm/TUI/Controller.hs [hunk 2]
```diff
@@ -1264,8 +1265,8 @@ tabComplete CompletionContext {..} names em theRepl = case theRepl ^. replPrompt
 
   possibleWords =
     names <> case ctxCreativeMode of
-      True -> reservedWords
-      False -> filter (`notElem` creativeWords) reservedWords
+      True -> S.toList reservedWords
+      False -> filter (`notElem` creativeWords) (S.toList reservedWords)
```
Comment (kostmo) [inline src/swarm-tui/Swarm/TUI/Controller.hs:1269]: This looks quadratic?  Convert `creativeWords` to a set and use `difference`

Comment (byorgey) [reply src/swarm-tui/Swarm/TUI/Controller.hs:1269]: Oops, good point.  I'll include this change with #1851.

**Pull request merged**